### PR TITLE
test: relay tests for TrySendTx/OnRecvPacket

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

This PR adds various `sdk.Msg` types to the existing relay tests and cleans up the structure.

This includes tx msgs from the following:
- x/staking
- x/gov/
- x/distribution
- IBC apps/transfer

closes: #510

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [x] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: modules/apps/27-interchain-accounts/keeper/relay_test.go [hunk 1]
```diff
@@ -28,45 +33,94 @@ func (suite *KeeperTestSuite) TestTrySendTx() {
...
-			}, true,
+			"success",
+			func() {
+				interchainAccountAddr, found := suite.chainB.GetSimApp().ICAKeeper.GetInterchainAccountAddress(suite.chainB.GetContext(), path.EndpointA.ChannelConfig.PortID)
+				suite.Require().True(found)
+
+				msg := &banktypes.MsgSend{
+					FromAddress: interchainAccountAddr,
+					ToAddress:   suite.chainB.SenderAccount.GetAddress().String(),
+					Amount:      sdk.NewCoins(sdk.NewCoin("stake", sdk.NewInt(100))),
```
Comment (seantking) [inline modules/apps/27-interchain-accounts/keeper/relay_test.go:44]:
```suggestion
					Amount:      sdk.NewCoins(sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(100))),
```

Comment (seantking) [reply modules/apps/27-interchain-accounts/keeper/relay_test.go:44]: Can do the same elsewhere in the test as well :)

Comment (damiannolan) [reply modules/apps/27-interchain-accounts/keeper/relay_test.go:44]: 100%, I only discovered this default var when reviewing @charleenfei's PR. Was going to update to align :)

Comment (damiannolan) [reply modules/apps/27-interchain-accounts/keeper/relay_test.go]: Done! 

I've also updated hardcoded strings `"channel-0"` and `"connection-0"` that are used throughout tests, with `ibctesting.FirstChannelID` and `ibctesting.FirstConnectionID` on my submodules branch.

File: modules/apps/27-interchain-accounts/keeper/relay_test.go [hunk 2]
```diff
@@ -129,105 +164,361 @@ func (suite *KeeperTestSuite) TestOnRecvPacket() {
...
 				packetData = icaPacketData.GetBytes()
-			}, false,
+			},
+			true,
 		},
 		{
-			"Invalid packet type", func() {
-				// build packet data
-				data, err := types.SerializeCosmosTx(suite.chainA.Codec, []sdk.Msg{&banktypes.MsgSend{}})
+			"interchain account fails to execute stakingtypes.MsgDelegate - insufficient funds",
```
Comment (seantking) [inline modules/apps/27-interchain-accounts/keeper/relay_test.go:216]: Just thinking out loud: do we need to test this? I imagine this is something that the SDK is covering.

File: modules/apps/27-interchain-accounts/keeper/relay_test.go [hunk 3]
```diff
@@ -129,105 +164,361 @@ func (suite *KeeperTestSuite) TestOnRecvPacket() {
...
-				amount, _ := sdk.ParseCoinsNormalized("100stake")
-				// Incorrect FromAddress
-				msg = &banktypes.MsgSend{FromAddress: suite.chainB.SenderAccount.GetAddress().String(), ToAddress: suite.chainB.SenderAccount.GetAddress().String(), Amount: amount}
-				// build packet data
-				data, err := types.SerializeCosmosTx(suite.chainA.Codec, []sdk.Msg{msg})
+			"unauthorised: interchain account not found for controller port ID",
+			func() {
+				path.EndpointA.ChannelConfig.PortID = "invalid-port-id"
+
+				data, err := types.SerializeCosmosTx(suite.chainA.GetSimApp().AppCodec(), []sdk.Msg{&banktypes.MsgSend{}})
```
Comment (seantking) [inline modules/apps/27-interchain-accounts/keeper/relay_test.go:460]: I think you need to populate the `&banktypes.MsgSend{}` in order to hit the intended error. I might be wrong though, but I'm guessing this fails on the message validation and not the ica auth.

Comment (seantking) [reply modules/apps/27-interchain-accounts/keeper/relay_test.go:460]: Ah sorry, I see you added a test below for that. The diff + unauthorized message threw me off.

File: modules/apps/27-interchain-accounts/keeper/relay_test.go [hunk 4]
```diff
@@ -129,105 +164,361 @@ func (suite *KeeperTestSuite) TestOnRecvPacket() {
...
 					Data: data,
 				}
+
 				packetData = icaPacketData.GetBytes()
-			}, false,
+			},
+			false,
+		},
+		{
+			"unauthorised: unexpected signer address",
```
Comment (seantking) [inline modules/apps/27-interchain-accounts/keeper/relay_test.go:473]:
```suggestion
			"unauthorised: signer address is not the interchain account associated with controller portID",
```

Comment (seantking) [reply modules/apps/27-interchain-accounts/keeper/relay_test.go:473]: just a nit but I preferred the explicit message :)

**Review by seantking — APPROVED** [review]

**Comment by seantking** [comment]
Super happy this is done :handshake: Nice work!

File: modules/apps/27-interchain-accounts/keeper/relay_test.go (hunk 4 above)
Comment (damiannolan) [reply modules/apps/27-interchain-accounts/keeper/relay_test.go:473]: Sure 👍

File: modules/apps/27-interchain-accounts/keeper/relay_test.go (hunk 2 above)
Comment (damiannolan) [reply modules/apps/27-interchain-accounts/keeper/relay_test.go:216]: Yeah it should probably be covered by the SDK I would imagine. I think we can leave it for now and remove later if needs be.

File: modules/apps/27-interchain-accounts/keeper/relay_test.go (hunk 3 above)
Comment (damiannolan) [reply modules/apps/27-interchain-accounts/keeper/relay_test.go:435]: Yep, these two tests cover two different errors returned from `AuthenticateTx` :)

**Review by colin-axner — APPROVED** [review]
Very clean, amazing work!

**Pull request merged**